### DISCUSSION TITLE
Using internal API route for static site generation

### DISCUSSION THREAD

**Discussion by niemiaszek** [vercel/next.js#47638] [Help] [answered]
### Summary

I'm developing my first nextjs project and I decided to start with nextjs 13 with App directory. I wanted to learn about ISR, SSR and SSG during this project. Repository with the link to the commit is attached.

I'm building exchange office web-page which main purpose is to present current exchange rates. These exchange rates change few times daily, so after first read I thought it's a good idea to use Static Side Generation with on-demand Incremental Static Regeneration when the exchange rates are updated. This way I wanted to present users light, static site.

As there are only few supported currencies in this office, I decided to store exchange rates (current and maybe archival) in private .json file. Obviously, as I'm complete web-dev noob, while implementing this page I made a rookie mistake. I set up API Route exposing JSON file contents and then fetching it in my wanted-to-be static site. This way in dev mode I was able to use my site (as it was SSR in dev mode), and I could also build it (it was built with cached fetch from dev SSR, which took me a long time to figure out when debuging). 

I was shocked when the repo wouldn't build when I tried to deploy it to Vercel, but I could run build in my local copy. Then I found [this Stack Overflow thread](https://stackoverflow.com/questions/66202840/fetch-error-when-building-next-js-static-website-in-production). I realized my mistake and changed the app not to use internal route in SSG, as it makes no sense to fetch non-existent internal route at build time. You can see my mistake in `app/page.tsx` and `pages/api/exchangetabledata.ts` in attached repo.

For most of you it might be expected behavior, that you can't fetch internal routes in SSG, but for me it made perfect sense. I had this imagination where Nextjs build the API route for me at the build time when it notices that i'm fetching internal API route. My imagination was partially real, because I could build SSG with catched fetch from internal API Route. I would expect next to run SSR first, and then to fetch and cache needed internal routes, and then to finally produce SSG with this cached fetch.

Maybe it is just my wild fantasy. What is your opinion on this feature and whole situation? I suspect that feature like this could promote some undesired practices or generate weird corner cases.

My request would be to make [fetching docs](https://beta.nextjs.org/docs/data-fetching/fundamentals) more informative on the topic of using internal routes (I'm missing something similar to Stack Overflow thread I mentioned before). It would make life easier for total noobs like me :smile:

PS:
I would be very glad if you had any comments or tips for current state of my web-page at github repository. Most crazy thing I'm developing now is Dashboard where I implemented authentication (next-auth with Github and Google + middleware) + whitelisted users (I store hashed user emails to compare with the ones given in authentication process). Dashboard will be the form, where the user will be able to input the new desired values of exchange rates, which will be posted to API route, updating the local json file with exchange rates and doing on-demand ISR for root page. Next step is setting up the server on the computer in the exchange office, which will make similar post request as the one from dashboard on every "ground truth" exchange rates update (local .xml file). This would reduce the need for the office employee to manually input the same data twice (once in the regular software handling the bookkeeping and then on web-page). I'm planning to setup github device flow authentication for this computer and prepare simple python script sending such request on every file update.

### Example

https://github.com/niemiaszek/exchange-office-webpage/tree/9785a6929ea9ee081f4ec3f55069134cd518eab3

**Comment by icyJoseph** [comment, marked as answer]
If I understand correctly, you were calling `api/something` during build time.

> This has been a developer pitfall since pages directory...

I guess the issue is that the framework shouldn't make any assumptions. Perhaps there's something running at `/api/something` during build time, it might be local server or whatever. I know it is a stretch but yeah. 

There's this bit of the docs, that doesn't say it really: https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes but it is a good rule of thumb.

And this goes into my bigger point, the whole point of GetServerSideProps, or React Server Components (RSC), is to fetch the data on the server, so you don't need to make an extra HTTP jump to an API route. 

The framework takes care of not leaking the server code to the client, https://next-code-elimination.vercel.app/ - in the case of RSC, the design of the entire rendering is made to prevent that. 

API routes exist to feed your clients, be it your own client side app, or 3rd parties that you wish to provide data to.

Perhaps something should be said, but every time people talk about RSC, they say how you can just get the Data you need already there, but out of a DRY instinct, or idk what else, we tend to group everything under API routes. 

> Whenever possible, we recommend fetching data inside [Server Components](https://beta.nextjs.org/docs/rendering/server-and-client-components). Server Components always fetch data on the server. This allows you to:
>
> - Have direct access to backend data resources (e.g. databases).

It ought to be said that, in the Nuxt framework you have to use a special $fetch function: https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls, which allows what you wanted to do.

Anyhow, good that you were able to fix it.

---

### A made up, exaggerated, counter example

An scenario where this would be unwanted: if you just wanted to pre-render a page. 

Let's imagine it works, and you were able to call API routes during build time, well, now the build output is:

- a static page with data from your database - non changing
- an API route that access your database

Now it is possible for bad actors to just keep on making requests to the API route, clogging your resources, when you didn't really needed it because your page is static. 

If we just connect to the DB that has this data from GetStaticProps, we would pre-render the page and destroy any possibility for bad actors to ever reach our DB.

If the data changed from time to time, we could activate ISR, or manually revalidate the page, still limiting how many requests actually arrive to our database.

**Reply by niemiaszek** [reply]
Thanks for your insights Joseph. Practical explanation you provided makes much sense and justifies the design choices of framework for me. Counter example like this what was somewhere in my mind. However, I think that this way it's harder to expose data, that could be shared with client both somewhere over pages and in API. If I wanted to share API with current API rates, I'd have to read that data both in SSG page and API Route (in counter to reading it once in API and then fetching it from here for page). Regarding these docs, I've focused on beta docs (as I'm using app directory), so I haven't seen those comments before. They were surely helpful - would be good to see something  similar in beta docs soon.
